jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs

```diff
@@ -9,32 +9,118 @@ mod error;
 mod openid_config;
 
 pub use error::KeyServiceError;
-use jsonwebtoken::jwk::JwkSet;
+use jsonwebtoken::jwk::Jwk;
 use jsonwebtoken::DecodingKey;
 pub(crate) use openid_config::*;
 use reqwest::blocking::Client;
+use serde::Deserialize;
 use std::collections::HashMap;
 use std::sync::Arc;
+use std::thread::sleep;
+use std::time::Duration;
 
+/// A wrapper around `reqwest::blocking::Client` providing HTTP request functionality with retry logic.
+///
+/// The `HttpClient` struct allows for sending GET requests with a retry mechanism that attempts to
+/// fetch the requested resource up to a maximum number of times if an error occurs.
+pub struct HttpClient {
+    client: reqwest::blocking::Client,
+    max_retries: u32,
+    retry_delay: Duration,
+}
+
+impl HttpClient {
+    pub fn default() -> Result<Self, KeyServiceError> {
+        let client = Client::builder()
+            .build()
+            .map_err(KeyServiceError::HttpClientInitialization)?;
+        Ok(Self {
+            client,
+            max_retries: 3,
+            retry_delay: Duration::from_secs(1),
+        })
+    }
+
+    #[cfg(test)]
+    pub fn fast_retry_delay() -> Result<Self, KeyServiceError> {
+        let client = Client::builder()
+            .build()
+            .map_err(KeyServiceError::HttpClientInitialization)?;
+        Ok(Self {
+            client,
+            max_retries: 3,
+            retry_delay: Duration::from_millis(10),
+        })
+    }
+
+    /// Sends a GET request to the specified URI with retry logic.
+    ///
+    /// This method will attempt to fetch the resource up to 3 times, with an increasing delay
+    /// between each attempt.
+    fn get(&self, uri: &str) -> Result<reqwest::blocking::Response, KeyServiceError> {
+        // Fetch the JWKS from the jwks_uri
+        let mut attempts = 0;
+        let response = loop {
+            match self.client.get(uri).send() {
+                Ok(response) => break response, // Exit loop on success
+                Err(e) if attempts < self.max_retries => {
+                    attempts += 1;
+                    // TODO: pass this message to the logger
+                    eprintln!(
+                        "Request failed (attempt {} of {}): {}. Retrying...",
+                        attempts, self.max_retries, e
+                    );
+                    sleep(self.retry_delay * attempts);
```

Comment:
How match time will take initialization if we have 10 trusted issuers and each return OK on 3th try?

Reply:
this will be changed to load lazily so they shouldn't be fetched all at the same time.